### ISSUE TITLE
lit-html is not working with shadyDOM

### ISSUE THREAD

**Issue by AndreasGalster** [Polymer/lit-html#136]
I'm sure you're aware of this already, just putting it on paper so we can track the future dev :). My lit-html component doesn't render styles properly if I'm turning shadyDOM on. 

# Expected result
All styles, custom properties and @apply mixins work as expected

# Actual result
None of them work properly

**Comment by daKmoR** [comment]
I guess that is not really a problem of shadyDom or shadyCSS as something like this will probably need to be done by the "app" or element itself.

litHtml does not "care" if it renders into a shadowRoot or some innerHtml.

You can see a simple demo here that uses shadowdom and shadycss.
https://github.com/daKmoR/lit-html-demos/blob/master/demo/wc02.html

**Comment by jamesisaac** [comment]
@daKmoR That example works to some degree, but from what I can tell, you're applying `ShadyCSS.prepareTemplate` to the throwaway `this.shadyTemplate`, which isn't used anywhere further. 
 So what this loses out on, is that the ShadyCSS classes (`style-scoped` and the name of the custom element) are not actually applied to all the elements in the subtree.  So, shadow DOM style encapsulation doesn't work, and the scoped styles ShadyCSS applied in the <head> don't do anything?

I had a go at applying `prepareTemplate` to lit-html's internal template, but I realise the fundamental flaw there is that lit-html is remembering the exact position of each TemplatePart, which will be broken if class names are inserted.  Seeing as when a TemplateResult is constructed, it doesn't really have insight into what element it's a part of (and therefore what classnames to add), I'm not sure a way round this.

**Comment by justinfagnani** [comment]
@sorvell is working on this

**Comment by tjmonsi** [comment]
I did something that sorts of hack this to make use of shadyCSS.

```
// in my connectedCallback() 

      if (window.ShadyCSS) {
        var template = document.createElement('template');
        render(this.template(), template.content);
        window.ShadyCSS.prepareTemplate(template, this.constructor.is);
        window.ShadyCSS.styleElement(/** @type {!HTMLElement} */(this));
      }

// and i override this component method
    template () {
      return html``;
    }
```

You guys can check my hacked polymer element-mixin where I removed the dom-module dependency and templating here: https://github.com/tjmonsi/tjmonsi-site/blob/master/src/core/mixins/element-mixin.js

**Comment by justinfagnani** [comment]
This was initially fixed a while back in #249. There are newer bugs and PRs for improved ShadyCSS integration.
